torchtune/utils/_generation.py

```diff
@@ -147,12 +150,24 @@ def generate(
                 [stop_token_mask, ~stop_token_reached.reshape(bsz, 1)], dim=-1
             )
 
+        # if incremental decoding is enabled, we can use the current position
+        # otherwise, we take the whole sequence up to the current position
+        if incremental_decoding:
+            curr_input_pos = input_pos[curr_pos].unsqueeze(0)
+        else:
+            curr_input_pos = input_pos[: curr_pos + 1]
+            tokens = generated_tokens.clone()
```

Comment:
I'm not entirely sure I need this clone...



cc someone who is better at pytorch than me

Reply:
I think it's safest to leave it since you need to keep track of tokens and generated tokens separately